detection of loops

memsharded opened the pull request:
This should solve issue https://github.com/conan-io/conan/issues/522

I have added the check in the graph node expansion through storage of ancestors. This is much cheaper than finding them on given graphs. Performance tests with 250 nodes graphs (both deep and breadth) show a very small impact on performance (1%).

annulen commented:
Out of curiosity, don't you consider using Python bindings of libsolv (or maybe hawkey which claims to provide higher-level API, not sure about it's current status though) to manage dependency graphs?

https://github.com/openSUSE/libsolv

memsharded commented:
Yes, we have consider several alternatives for managing the deps graph. In fact, we had in previous version a python library for graphs, and finally we removed that dependency. 

- They are difficult to integrate/maintain in all systems conan is running. Bindings, as we understand them, are just a layer over the native library, which has to be managed, ensure that it builds in different OSs, packaged in the installers, packaged in PyPi... that is a lot of work, and from our experience they come with a lot of maintenance burden and troubleshooting. 
- Pure python libraries of course are much easier to integrate. The reason we removed the library we were using was that the functionality we need is pretty simple, the implementation is very straightforward. When using that library we had very bad performance, which was easily solved by caching the neighbors of each node. And we realized that it was less work to debug and maintain our own graph model than workarounding a generic purpose one. The current implementation of the DepsGraph has approx 200 lines, and that is because it contains some of the business logic.
- In any case, performance over computation of the graph is not that critical (with our implementation), and it takes a very small percentage of the processing time of conan. 
- Using the knowledge of the process allows better performance. As in this case. If a generic loop detection algorithm had to be run over a given graph, at each step of the graph construction, that could have an impact. Integrating the loop detection as I did is very simple and performant.

annulen commented:
Just letting you know that recent tendency amongst Linux package managers is moving to libsolv (SAT solver), it's very fast and algorithm is reliable
